analyzer/src/steps/typing.rs

```diff
@@ -166,6 +164,163 @@ fn ascribe_literal(lit: &Literal) -> TypedExpr {
     }
 }
 
+fn ascribe_template_string(
+    tpl: &TemplateString,
+    exploration: &mut Exploration,
+    relations: &Relations,
+    diagnostics: &mut Vec<Diagnostic>,
+    env: &Environment,
+    state: TypingState,
+) -> TypedExpr {
+    if tpl.parts.is_empty() {
+        return TypedExpr {
+            kind: ExprKind::Literal(LiteralValue::String(String::new())),
+            ty: STRING,
+            segment: tpl.segment(),
+        };
+    }
+    let plus_method = exploration
+        .engine
+        .get_method_exact(
+            STRING,
+            name_operator_method(BinaryOperator::Plus),
+            &[STRING],
+            STRING,
+        )
+        .expect("string type should have a concatenation method")
+        .definition;
+    let mut it = tpl.parts.iter().map(|part| {
+        let typed_part = ascribe_types(
+            exploration,
+            relations,
+            diagnostics,
+            env,
+            part,
+            state.without_local_type(),
+        );
+        convert_into_string(typed_part, exploration, diagnostics, state)
+    });
+    let acc = it.next().unwrap();
+    it.fold(acc, |acc, current| {
+        let segment = current.segment.clone();
+        TypedExpr {
+            kind: ExprKind::MethodCall(MethodCall {
+                callee: Box::new(acc),
+                arguments: vec![current],
+                definition: plus_method,
+            }),
+            ty: STRING,
+            segment,
+        }
+    })
+}
+
+fn ascribe_assign(
+    assign: &Assign,
+    exploration: &mut Exploration,
+    relations: &Relations,
+    diagnostics: &mut Vec<Diagnostic>,
+    env: &Environment,
+    state: TypingState,
+) -> TypedExpr {
+    let rhs = ascribe_types(
+        exploration,
+        relations,
+        diagnostics,
+        env,
+        &assign.value,
+        state.with_local_type(),
+    );
+    let symbol = env.get_raw_symbol(assign.segment()).unwrap();
+    let actual_type = exploration
+        .get_type(
+            exploration
+                .ctx
+                .get(relations, state.source, symbol)
+                .unwrap()
+                .type_id,
+        )
+        .unwrap();
+    if actual_type.is_callable() {
+        diagnostics.push(
+            Diagnostic::new(
+                DiagnosticID::TypeMismatch,
+                state.source,
+                format!("Cannot treat function `{}` as a variable", assign.name),
```

Comment:
this is false, you can access functions as variables, but you can't _assign_ them
```suggestion
                format!("Cannot treat functions as variables for assignement", assign.name),
```
or
```suggestion
                format!("functions cannot be assigned", assign.name),
```

Reply:
Functions are not variables, even outside assignments. They can be assigned to a variable, but are not themselves variables.

Reply:
Renamed to convey that named types are different from variables.